packages/wonder-blocks-modal/src/components/flexible-dialog.tsx

```diff
@@ -0,0 +1,225 @@
+import * as React from "react";
+import {StyleSheet} from "aphrodite";
+import type {PropsFor, StyleType} from "@khanacademy/wonder-blocks-core";
+
+import {Id, View} from "@khanacademy/wonder-blocks-core";
+import {breakpoint} from "@khanacademy/wonder-blocks-tokens";
+import {Heading} from "@khanacademy/wonder-blocks-typography";
+import ModalDialog from "./modal-dialog";
+import FlexiblePanel, {BackgroundStyles} from "./flexible-panel";
+
+/**
+ * The visible title content for the dialog. An ID will be automatically applied
+ * for aria-labelledby on the dialog to the heading content or a div wrapper.
+ */
+type TitleAndAriaLabelledBy = {
+    title: React.ReactNode;
+    "aria-label"?: never;
+    "aria-labelledby"?: string;
+};
+/**
+ * The optional accessible name of the modal.
+ * This is useful when there is no main heading in the dialog. A visible
+ * heading with an id is preferred for automatic matching with aria-labelledby.
+ */
+type AriaLabelOnly = {
+    title?: never;
+    "aria-label": string;
+    "aria-labelledby"?: never;
+};
+/**
+ * The optional ID of the content describing this dialog, if applicable.
+ */
+type AriaLabelledByOnly = {
+    title?: never;
+    "aria-label"?: never;
+    "aria-labelledby": string;
+};
+
+/**
+ * One of the labeling methods is required:
+ */
+type AccessibleLabelProps =
+    | TitleAndAriaLabelledBy
+    | AriaLabelOnly
+    | AriaLabelledByOnly;
+
+type Props = AccessibleLabelProps & {
+    /**
+     * The main heading of the modal, for labeling the dialog.
+     * If string content, an ID will be generated for the heading for aria-labelledby on the dialog.
+     * If a node, an ID will be applied to a DIV wrapping the node.
+     */
+    title?: React.ReactNode | string;
+    /**
+     * An optional id parameter for the main heading. If one is not provided,
+     * an ID will be generated.
+     */
+    titleId?: string;
+    /**
+     * The content of the modal. Supports a render prop for placing the title in a slot.
+     */
```

Comment:
I think it would be worth mentioning in the description that in addition to supporting a render prop, a `FlexiblePanel` should be used with `FlexibleDialog`!

I found that TS can't make children components type safe at this time so `React.ReactElement<PropsFor<typeof FlexiblePanel>>` doesn't enforce the content element to be a `FlexiblePanel` component 🥲 ([Related Slack thread](https://khanacademy.slack.com/archives/C028T2HDX3P/p1741210473344549))

Reply:
The `FlexiblePanel` is used internally, so it wouldn't be something for consumers. Is it still worth adding a comment?

Reply:
Ah right, good point! We don't need to mention it in the comment :) 

Should the type be `React.ReactElement` only without the `PropsFor<typeof FlexiblePanel>` part to avoid confusion?

Reply:
Also, related question! I see the render function returns a `React.ReactNode` element. Do you think it should be the same type as when a render function is not used? (ie. Should they both be `ReactElement` or `ReactNode`? It looks like ReactNode supports more types!) 

I'm still figuring out when to use which type, so I'm curious what others think about this!